File: Machine Learning A-Z - Hands-On Python & R Data Science/XGBoost/xgboost_intro.R
# Extreme Gradient Boosting (XGBoost)

# Importing the dataset
dataset = read.csv('Data\\Churn_Modelling.csv')
dataset = dataset[4:ncol(dataset)]

# Encoding categorical variables as factors
dataset$Geography = as.numeric(factor(x = dataset$Geography,
                                      levels = c('France', 'Spain', 'Germany'),
                                      labels = c(1, 2, 3)))
dataset$Gender = as.numeric(factor(x = dataset$Gender,
                                   levels = c('Female', 'Male'),
                                   labels = c(0, 1)))

# Splitting the dataset into the Training set and Test set
library(caTools)
set.seed(123)
split = sample.split(Y = dataset$Exited, SplitRatio = 0.8)
training_set = subset(x = dataset, split == TRUE)
test_set = subset(x = dataset, split == FALSE)

# Feature Scaling
training_set[, -11] = scale(training_set[, -11])
test_set[, -11] = scale(test_set[, -11])

# Fitting XGBoost to the Training Set
library(xgboost)
classifier = xgboost(data = as.matrix(training_set[-11]),
                     label = training_set$Exited,
                     nrounds = 10)

# Predicting the Test Set results
y_prob = predict(classifier, newdata = as.matrix(test_set[-11]))
y_pred = ifelse(y_prob >= 0.5, 1, 0)

# Make the Confusion Matrix
cm = table(test_set[, 11], y_pred)

# Applying k-Fold Cross Validation
library(caret)
folds = createFolds(training_set$Exited, k = 10)
cv = lapply(X = folds, FUN = function(x) {
  training_fold = training_set[-x, ]
  test_fold = training_set[x, ]
  classifier = xgboost(data = as.matrix(training_set[-11]),
                       label = training_set$Exited,
                       nrounds = 10)
  y_prob = predict(classifier, newdata = as.matrix(test_fold[-11]))
  y_pred = ifelse(y_prob >= 0.5, 1, 0)
  cm = table(test_fold[, 11], y_pred)
  accuracies = (cm[1, 1] + cm[2, 2]) / (cm[1, 1] + cm[2, 2] + cm[1, 2] + cm[2, 1])
  return(accuracies)
})
mean(as.numeric(cv))

File: Machine Learning A-Z - Hands-On Python & R Data Science/XGBoost/xgboost_intro.py
# Extreme Gradient Boosting (XGBoost) - Introduction

# # Importing the Libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Data\\Churn_Modelling.csv')
X = dataset.iloc[:, 3:-1].values
y = dataset.iloc[:, -1].values

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
le_X_1 = LabelEncoder()
X[:, 1] = le_X_1.fit_transform(X[:, 1])
le_X_2 = LabelEncoder()
X[:, 2] = le_X_2.fit_transform(X[:, 2])
ohe = OneHotEncoder(categorical_features=[1])
X = ohe.fit_transform(X).toarray()
X = X[:, 1:]

# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    test_size=0.2,
                                                    random_state=0)

# Fitting XGBoost to the Training Set
from xgboost import XGBClassifier
classifier = XGBClassifier()
classifier.fit(X_train, y_train)

# Predicting the Test Set results
y_pred = classifier.predict(X_test)


# Making the confusion matrix and finding accuracy
import seaborn as sns
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
sns.heatmap(cm, annot=True)
accuracy_score(y_test, y_pred)

# Applying k-Fold Cross Validation
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier,
                             X = X_train,
                             y = y_train,
                             cv = 10)
accuracies.mean(), accuracies.std()
